Drop all CIB replace notify code

nrwahl2 opened the pull request:
Every time a replace notification is sent, a diff notification is also sent. So the replace notification is redundant.

nrwahl2 commented on lib/cib/cib_utils.c:
```diff
@@ -77,6 +78,116 @@ cib_diff_version_details(xmlNode * diff, int *admin_epoch, int *epoch, int *upda
+ *       supported value for \p element.
+ */
+bool
+cib__element_in_patchset(const xmlNode *msg, const char *element)
```
I haven't decided what approach I'm going to use when we want to do things soon like "check whether the status section OTHER THAN transient attributes changed."

I might convert this to `cib__xpath_in_patchset()` (or create a new such function) that does regex matching. It may or may not be able to support XPaths with multiple terms (e.g., `/path1|/path2`).

nrwahl2 commented:
Hold off on looking at this beyond maybe a skim.

This craps out pretty badly (controller never gets any membership data), so I'll have to see which one of the commits breaks it first. I'm guessing "Drop CIB replace call recording and forgetting."

nrwahl2 commented on daemons/controld/controld_join_dc.c:
```diff
@@ -821,6 +822,7 @@ do_dc_join_ack(long long action,
     }
     free(xpath);
     free(join_from);
+    free(xpath);
```
leftover that got auto-applied during rebase, fixing

kgaillot commented on daemons/attrd/attrd_cib.c:
```diff
@@ -45,21 +45,16 @@ attrd_cib_destroy_cb(gpointer user_data)
 
-    if (attrd_shutting_down(true)) {
+    if (attrd_shutting_down(false)) {
+        // @TODO Pass true? attrd_cib_replaced_cb() passed true.
```
Definitely true (see a1a9c54cfc). If we wanted to preserve existing behavior, we could check it with false before checking alerts and then call it again with true before checking nodes. However I think it's reasonable to ignore any changes to alerts if we're intending to shut down.

kgaillot commented on daemons/attrd/attrd_cib.c:
```diff
@@ -73,25 +68,14 @@ attrd_cib_replaced_cb(const char *event, xmlNode *msg)
-                           cib_change_section_nodes
-                           |cib_change_section_status)) {
+    if (cib__element_in_patchset(msg, XML_CIB_TAG_NODES)
+        || cib__transient_attrs_in_patchset(msg)) {
```
This is more than a refactor since we're changing what we're looking for -- worth calling out in commit message or even its own commit

nrwahl2 replied:
Definitely. I must've been messing around and forgotten to separate it. One commit should keep all the logic the same while dropping the replace callback.

kgaillot reviewed (commented):
This is so much cleaner and easier to follow. Not sure why replace callbacks were ever used.

nrwahl2 commented on lib/cib/cib_utils.c:
```diff
@@ -77,6 +78,259 @@ cib_diff_version_details(xmlNode * diff, int *admin_epoch, int *epoch, int *upda
+    }
+}
+
+#define REGEX_ID            "\\[@" XML_ATTR_ID "='[^']*'\\]"
```
Unfortunately for v2 patchsets, we can't use XPath queries. The best we can do is regex matching against the `path` attribute of each `<change>` element. The path will have `.../node_state[@id='1']` or similar.

Having to do string matching makes all of this clunkier than it otherwise could be.

nrwahl2 commented on daemons/controld/controld_cib.c:
```diff
@@ -134,7 +134,9 @@ handle_cib_disconnect(gpointer user_data)
 {
-    if (pcmk__alert_in_patchset(msg, TRUE)) {
+    if (cib__element_in_patchset(msg, XML_CIB_TAG_ALERTS)
+        || cib__element_in_patchset(msg, XML_CIB_TAG_CRMCONFIG)) {
```
Agreed, I figure the readability is worth the little bit of extra work to iterate a second time.

nrwahl2 commented on daemons/attrd/attrd_cib.c:
```diff
@@ -45,21 +45,16 @@ attrd_cib_destroy_cb(gpointer user_data)
 
 static void
-attrd_cib_replaced_cb(const char *event, xmlNode *msg)
+attrd_cib_updated_cb(const char *event, xmlNode *msg)
```
Not fully, this is still in draft state. When I left it, it definitely had bugs.

nrwahl2 commented:
Updated commits to address review. For now, I removed the parts that add checks more granular than `XML_CIB_TAG_STATUS`. Haven't run `cts-lab` yet.

I just realized this still won't be quite right. I'm checking `cib__op_attr_replaces`, but transactions replace the CIB. Our one CIB transaction in the controller will hit the "replaces" condition in `attrd` and trigger a write.

I doubt it's any worse than the current situation though. The controller's CIB transaction should already be triggering a replace notification. `attrd` currently uses `modify` rather than a transaction, so that doesn't trigger a loop of replace callbacks in response to the controller.

nrwahl2 commented on daemons/controld/controld_cib.c:
```diff
@@ -53,45 +53,46 @@ handle_cib_disconnect(gpointer user_data)
+            || cib__element_in_patchset(msg, XML_CIB_TAG_STATUS))) {
 
-        /* start the join process again so we get everyone's LRM status */
+        /* Nodes or status section was replaced. We assume this was triggered by
```
comment is not quite right since transactions trigger replacements. likewise for attrd

nrwahl2 commented:
cts-lab passed. Need to update a couple of comments and then this is probably ready.

nrwahl2 commented on lib/cib/cib_utils.c:
```diff
@@ -77,6 +78,128 @@ cib_diff_version_details(xmlNode * diff, int *admin_epoch, int *epoch, int *upda
+                return true;
+            }
+
+        } else if (pcmk__starts_with(diff_xpath, element_xpath)) {
```
On second thought, this probably should be an `if` rather than an `else if`.

Suppose we're concerned about `/elem1`. Suppose we have a create operation where `/elem1/elem2/elem3` is created (as a child of `/elem1/elem2`. I don't think we'd hit any of these conditions. But `/elem1` did change.

nrwahl2 commented:
Updated with only slight changes. Running another cts-lab currently as a sanity check. Flipping to ready.

nrwahl2 commented:
retest this please

nrwahl2 commented:
Removed three commits that were unrelated to the others ones in the first place. They'll be slightly reworked for a future PR.

kgaillot commented on include/crm/common/Makefile.am:
```diff
@@ -47,6 +46,7 @@ noinst_HEADERS = acl_internal.h 	\
 				 health_internal.h	\
 				 internal.h		\
 				 io_internal.h		\
+				 ipc_attrd_internal.h	\
```
You can drop this commit too, @clumens has a PR changing this to a wildcard

nrwahl2 replied:
Done

kgaillot reviewed (commented):
looks good

nrwahl2 commented on lib/cib/cib_utils.c:
```diff
@@ -77,6 +78,259 @@ cib_diff_version_details(xmlNode * diff, int *admin_epoch, int *epoch, int *upda
+    patchset = get_message_xml(msg, F_CIB_UPDATE_RESULT);
+    CRM_CHECK(patchset != NULL, return false);
+
+    crm_element_value_int(patchset, "format", &format);
```
Interesting. `format` is going away soon, so I'm not going to worry about that part. However, I think it makes more sense for this function to accept a patchset than to accept the whole diff notification message. That would also make a future planned change easier.

The caller can fetch the patchset and can check the CIB return code if desired

nrwahl2 commented:
just updated copyright in one file

kgaillot commented:
rebase on main to resolve conflicts

nrwahl2 commented:
Rebased on main and `s/cib__diff_notify_patchset/cib__get_notify_patchset/g` as I'd done in the other PR

kgaillot commented on lib/cib/cib_utils.c:
```diff
@@ -78,6 +78,43 @@ cib_diff_version_details(xmlNode * diff, int *admin_epoch, int *epoch, int *upda
+    }
+
+    if ((crm_element_value_int(msg, F_CIB_RC, &rc) != 0) || (rc != pcmk_ok)) {
+        crm_warn("Ignore failed CIB update: %s (%d)", pcmk_strerror(rc), rc);
```
while we're here, the preferred style these days is `"... %s " CRM_XS " rc=%d"` so system log is more user-friendly

kgaillot reviewed (approved):
looks good
